niworkflows/interfaces/utils.py

```diff
@@ -256,3 +257,150 @@ def _gen_reference(fixed_image, moving_image, fov_mask=None, out_file=None,
+    output_spec = SanitizeImageOutputSpec
+
+    def _run_interface(self, runtime):
+        img = nb.load(self.inputs.in_file)
+        out_report = os.path.join(runtime.cwd, 'report.html')
+
+        # Retrieve xform codes
+        sform_code = int(img.header._structarr['sform_code'])
+        qform_code = int(img.header._structarr['qform_code'])
+
+        # Check qform is valid
+        valid_qform = False
+        try:
+            img.get_qform()
+            valid_qform = True
+        except ValueError:
+            pass
+
+        # Matching affines
+        matching_affines = valid_qform and np.allclose(img.get_qform(), img.get_sform())
+
+        save_file = False
+
+        # Both match, qform valid (implicit with match), codes okay -> do nothing, empty report
+        if matching_affines and qform_code > 0 and sform_code > 0:
+            self._results['out_file'] = self.inputs.in_file
+            open(out_report, 'w').close()
+
+        # Row 2:
+        elif valid_qform and qform_code > 0:
+            img.set_sform(img.get_qform(), qform_code)
+            save_file = True
+            warning_txt = 'Note on orientation: sform matrix set'
+            description = """\
+<p class="elem-desc">The sform has been copied from qform.</p>
+"""
+        # Rows 3-4:
+        # Note: if qform is not valid, matching_affines is False
+        elif sform_code > 0 and (not matching_affines or qform_code == 0):
+            img.set_qform(img.get_sform(), sform_code)
+            save_file = True
+            warning_txt = 'Note on orientation: qform matrix overwritten'
+            description = """\
+<p class="elem-desc">The qform has been copied from sform.</p>
+"""
+            if not valid_qform and qform_code > 0:
+                warning_txt = 'WARNING - Invalid qform information'
+                description = """\
+<p class="elem-desc">
+    The qform matrix found in the file header is invalid.
+    The qform has been copied from sform.
+    Checking the original qform information from the data produced
+    by the scanner is advised.
+</p>
+"""
+        # Rows 5-6:
+        else:
+            affine = img.affine
+            img.set_sform(affine, nb.nifti1.xform_codes['scanner'])
+            img.set_qform(affine, nb.nifti1.xform_codes['scanner'])
+            save_file = True
+            warning_txt = 'WARNING - Missing orientation information'
+            description = """\
+<p class="elem-desc">
+    Orientation information could not be retrieved from the image header.
+    The qform and sform matrices have been set to a default, LAS-oriented affine.
+    Analyses of this dataset MAY BE INVALID.
+</p>
+"""
+
+        if (self.inputs.force_float32 and np.dtype(img.get_data_dtype()).itemsize > 4) or self.inputs.n_volumes_to_discard:
+            in_data = img.get_data()
+
+            if (self.inputs.force_float32 and np.dtype(img.get_data_dtype()).itemsize > 4):
+                in_data = in_data.astype(np.float32)
```

Comment:
How about replacing the previous four lines with:


```Python
in_data = img.get_fdata(dtype=np.float32) if self.inputs.force_float32 else img.dataobj
```

For large datasets, this could save us some memory.

This bumps our nibabel dependency up to 2.2.0, but we're going to need that if we want to start doing CIFTI outputs, anyway.

Reply:
`dataobj` is cool when you don't need to read all of the data, but since we need save the whole file would there be any gains?

Reply:
Sure, you can use `get_data()`. The only advantage to `dataobj` here is that if you're dropping some volumes (which is the case when you aren't forcing float), you don't have to load them into memory first. It's going to be a pretty small savings, so I don't feel strongly here.

The main point was the `get_fdata` which lets nibabel do the work of coercing the type while keeping memory usage under control.

```Python
# Loads full file and fills cache
in_data = img.get_data()
# Makes new array, old one still around until `img` is garbage collected
in_data = in_data.astype(np.float32)
```